src/app/MessageDef.cpp

```diff
@@ -0,0 +1,3505 @@
+    {                                                                                                                              \
+        gPrettyPrintingDepthLevel--;                                                                                               \
+    } while (0)
+
+static void PrettyPrintIM(bool aIsNewLine, const char * aFmt, ...)
+{
+    va_list args;
+    size_t ret;
+    size_t sizeLeft;
+
+    va_start(args, aFmt);
+
+    if (aIsNewLine)
+    {
+        if (gCurLineBufferSize)
+        {
+            // Don't need to explicitly NULL-terminate the string because
+            // snprintf takes care of that.
+            ChipLogDetail(DataManagement, "%s", gLineBuffer);
+            gCurLineBufferSize = 0;
+        }
+
+        for (uint32_t i = 0; i < gPrettyPrintingDepthLevel; i++)
+        {
+            if (sizeof(gLineBuffer) > gCurLineBufferSize)
+            {
+                sizeLeft = sizeof(gLineBuffer) - gCurLineBufferSize;
+                ret      = (size_t)(snprintf(gLineBuffer + gCurLineBufferSize, sizeLeft, "\t"));
+                if (ret > 0)
+                {
+                    gCurLineBufferSize += std::min(ret, sizeLeft);
+                }
+            }
+        }
+    }
+
+    if (sizeof(gLineBuffer) > gCurLineBufferSize)
+    {
+        sizeLeft = sizeof(gLineBuffer) - gCurLineBufferSize;
+        ret      = (size_t)(vsnprintf(gLineBuffer + gCurLineBufferSize, sizeLeft, aFmt, args));
+        if (ret > 0)
+        {
+            gCurLineBufferSize += std::min(ret, sizeLeft);
+        }
+    }
+
+    va_end(args);
+}
+#else // CHIP_DETAIL_LOGGING
+#define PRETTY_PRINT_CHECKPOINT()
+#define PRETTY_PRINT(fmt, ...)
+#define PRETTY_PRINT(fmt, ...)
+#define PRETTY_PRINT_INCDEPTH()
+#define PRETTY_PRINT_DECDEPTH()
+#endif // CHIP_DETAIL_LOGGING
+
+Parser::Parser() : mOuterContainerType(chip::TLV::kTLVType_NotSpecified) {}
+
+void Parser::Init(const chip::TLV::TLVReader & aReader, chip::TLV::TLVType aOuterContainerType)
+{
+    mReader.Init(aReader);
+    mOuterContainerType = aOuterContainerType;
+}
+
+CHIP_ERROR Parser::GetReaderOnTag(const uint64_t aTagToFind, chip::TLV::TLVReader * const apReader) const
+{
+    return mReader.FindElementWithTag(aTagToFind, *apReader);
+}
+
+template <typename T>
+CHIP_ERROR Parser::GetUnsignedInteger(const uint8_t aContextTag, T * const apLValue) const
+{
+    return GetSimpleValue(aContextTag, chip::TLV::kTLVType_UnsignedInteger, apLValue);
+}
+
+template <typename T>
+CHIP_ERROR Parser::GetSimpleValue(const uint8_t aContextTag, const chip::TLV::TLVType aTLVType, T * const apLValue) const
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::TLV::TLVReader reader;
+
+    *apLValue = 0;
+
+    err = mReader.FindElementWithTag(chip::TLV::ContextTag(aContextTag), reader);
+    SuccessOrExit(err);
+
+    VerifyOrExit(aTLVType == reader.GetType(), err = CHIP_ERROR_WRONG_TLV_TYPE);
+
+    err = reader.Get(*apLValue);
```

Comment:
So.. this is completely unsafe.  It will do weird and broken things if the value in the TLV is out of range for type T.

Reply:
https://github.com/project-chip/connectedhomeip/issues/3957